### PR TITLE
Add support for creation and execution of temporary views

### LINKED ISSUE #21: Create and execute a temporary view

### PR DESCRIPTION
**Pull request by ermyas**
Fixes #21

### CODE REVIEW

File: src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala [hunk 1]
```diff
@@ -31,9 +31,10 @@ trait QueryStrategy {
   }
 
   def queryByIds[K: W, Q: R](client: Client, db: String, url: String, ids: Seq[K], ps: Map[String, String]): Task[Q] = {
-    if (ids.isEmpty)
-      Res.Error("not_found", "No IDs specified").toTask[Q]
-    else
-      client.post[Req.DocKeys[K], Q](url, Status.Ok, Req.DocKeys[K](ids), ps.toSeq)
+    query[Req.DocKeys[K], Q](client, db, url, Req.DocKeys[K](ids), ps)
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala]: `Req.DocKeys(ids)` - the `K` type parameter is not necessary, it gets inferred.

File: src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala [hunk 2]
```diff
@@ -31,9 +31,10 @@ trait QueryStrategy {
   }
 
   def queryByIds[K: W, Q: R](client: Client, db: String, url: String, ids: Seq[K], ps: Map[String, String]): Task[Q] = {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala]: Should we decide on the maximum line length? This seems too long.

File: src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala [hunk 3]
```diff
@@ -31,9 +31,10 @@ trait QueryStrategy {
...
-    if (ids.isEmpty)
-      Res.Error("not_found", "No IDs specified").toTask[Q]
-    else
-      client.post[Req.DocKeys[K], Q](url, Status.Ok, Req.DocKeys[K](ids), ps.toSeq)
+    query[Req.DocKeys[K], Q](client, db, url, Req.DocKeys[K](ids), ps)
+  }
+
+  def query[B: W, Q: R](client: Client, db: String, url: String, body: B, ps: Map[String, String]): Task[Q] = {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala]: Could you please give this method a bit more descriptive name? Maybe postQuery?

File: src/main/scala/com/ibm/couchdb/Req.scala [hunk 4]
```diff
@@ -22,4 +22,5 @@ object Req {
 
   case class DocKeys[K](keys: Seq[K])
 
+  case class DocViewWithKeys[K](keys: Seq[K], view: CouchView)
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/Req.scala]: Call this just ViewWithKeys?

File: src/main/scala/com/ibm/couchdb/api/builders/ViewQueryBuilder.scala [hunk 5]
```diff
@@ -133,11 +131,37 @@ case class ViewQueryBuilder[K, V](client: Client,
...
+    case None => query[Q](client, db, url, ps)
   }
 
   private def queryByIds[Q: R](ids: Seq[K], ps: Map[String, String]): Task[Q] = {
-    queryByIds[K, Q](client, db, s"/$db/_design/$design/_view/$view", ids, ps)
+    if (ids.isEmpty)
+      Res.Error("not_found", "No IDs specified").toTask
+    else {
+      temporaryView match {
+        case Some(t) => query[DocViewWithKeys[K], Q](client, db, url, DocViewWithKeys(keys = ids, t), ps)
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/ViewQueryBuilder.scala]: I like to qualify the request types with `Req.`, like `Req.DocViewWithKeys` -- this makes it clearer that it's a part of a request.

File: src/test/scala/com/ibm/couchdb/api/QueryTemporaryViewSpec.scala [hunk 6]
```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2015 IBM Corporation
```
Comment (beloglazov) [inline src/test/scala/com/ibm/couchdb/api/QueryTemporaryViewSpec.scala]: 2016

File: src/test/scala/com/ibm/couchdb/spec/Fixtures.scala [hunk 7]
```diff
@@ -56,30 +87,10 @@ trait Fixtures {
...
-        |}
-        """.stripMargin),
-      FixViews.compound → CouchView(map =
-        """
-          |function(doc) {
-          | var d = doc.doc;
-          | emit([d.age, d.name], d);
-          |}
-        """.stripMargin)),
+                 FixViews.names → FixViews.namesView,
```
Comment (beloglazov) [inline src/test/scala/com/ibm/couchdb/spec/Fixtures.scala]: Just a minor thing, could you please replace → with -> for consistency with other places? :)

File: src/main/scala/com/ibm/couchdb/api/builders/ViewQueryBuilder.scala [hunk 8]
```diff
@@ -133,11 +131,37 @@ case class ViewQueryBuilder[K, V](client: Client,
...
+        case None => queryByIds[K, Q](client, db, url, ids, ps)
+      }
+    }
+  }
+  private def url: String = (view, design) match {
```
Comment (beloglazov) [inline src/main/scala/com/ibm/couchdb/api/builders/ViewQueryBuilder.scala:149]: Blank line between methods.

File: src/test/scala/com/ibm/couchdb/api/QueryTemporaryViewSpec.scala [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ibm.couchdb.api
+
+import com.ibm.couchdb.spec.CouchDbSpecification
+
+class QueryTemporaryViewSpec extends CouchDbSpecification {
```
Comment (beloglazov) [inline src/test/scala/com/ibm/couchdb/api/QueryTemporaryViewSpec.scala:21]: I hope this won't diverge too much with QueryViewSpec :)

File: src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala (hunk 2 above)
Comment (ermyas) [inline src/main/scala/com/ibm/couchdb/api/builders/QueryStrategy.scala]: my settings are at 120 chars, this is just within that limit. But I can restructure this

File: src/test/scala/com/ibm/couchdb/api/QueryTemporaryViewSpec.scala (hunk 6 above)
Comment (ermyas) [inline src/test/scala/com/ibm/couchdb/api/QueryTemporaryViewSpec.scala]: Fixed my header auto-generator so this should not be an issue again :-)

**Pull request merged**